src/common/TreeOps.h

```diff
@@ -255,4 +259,30 @@ getAtoms(PTRef tr, Logic & logic)
     return atoms;
 }
 
+template<typename TPred>
+class TermCollectorConfig : public DefaultVisitorConfig {
+    TPred predicate;
+    vec<PTRef> gatheredTerms;
+public:
+    TermCollectorConfig(TPred predicate) : predicate(std::move(predicate)) {}
```

Comment:
```suggestion
    TermCollectorConfig(TPred && predicate) : predicate(std::move(predicate)) {}
```

Reply:
I prefer passing the predicate by value.
Here is some discussion on this topic: https://stackoverflow.com/questions/48803927/why-does-stdfind-iffirst-last-p-not-take-predicate-by-reference

Reply:
> > Can we convert also getAtoms in TreeOps.h?
> 
> Probably yes, although it seems to have some non-trivial logic, does not examining terms beyond children of `BooleanOperator`. We would have to extend `TermCollectorConfig` to accept override for `previsit` as well, I guess... Seems a bit more complicated than what I wanted to achieve at the moment.

Maybe so...  Do you know if it is currently used outside opensmt?  The only use within opensmt is in a unit test.

> 
> I am also thinking of having a general method that wouldn't collect the subterms, but it would instead accept an action that would be executed on every subterm (satisfying given predicate). Then we could maybe avoid **collecting** subterms altogether.

Ok, sounds like defining the semantics of that operation is not trivial (e.g., can it change the tree?), but also good if we can make it work for something.

Reply:
Interesting 🤔

Ok, let's go with the value.